src/fiat/drnvtx/dr_nvtx_map.cc

```diff
@@ -0,0 +1,78 @@
+#include <unordered_map>
+#include <cstring>
+#include <iostream>
+
+#include "dr_nvtx_map.h"
+
+using namespace std;
+
+extern "C" double MPI_Wtime ();
+#pragma weak MPI_Wtime
+
+namespace
+{
+  struct counter 
+  {
+    int calls = 0;
+    double elapsed = 0;
+    double t0 = 0;
+  };
+
+  template <class _Tp>
+  struct equal_to : public binary_function<_Tp, _Tp, bool>
+  {
+    bool operator()(const _Tp& __x, const _Tp& __y) const
+    {  
+      return strcmp( __x, __y ) == 0; 
+    }
+  };
+
+
+  struct hash
+  {
+    //BKDR hash algorithm
+    int operator() (const char * str) const
+    {
+      int seed = 131;//31  131 1313 13131131313 etc//
+      int hash = 0;
+      while(*str)
+      {
+          hash = (hash * seed) + (*str);
+          str ++;
+      }
+
+      return hash & (0x7FFFFFFF);
+    }
+  };
+
+  counter * stack[128];
+
+  std::unordered_map<char const*, counter, hash,  equal_to<const char*>> map;
+
+  int ilast = 0;
+};
+
+extern "C" int dr_nvtx_map_start (const char * str) 
+{
+  counter * elem = &(map[str]);
+  ilast++;
+  stack[ilast] = elem;
+  elem->calls ++;
+  if (elem->calls >= 11 && elem->elapsed < 0.0001) 
```

Comment:
Hi @pmarguinaud, I've been doing some work on this patch to integrate it into the core of drhook so that it's faster, works with all support drhook languages, and is easier for us to maintain.

This condition is giving me a bit of confusion though. My understanding is that it's meant to eliminate high call but low cumulative runtime drhook regions from the nvtx trace, so as to reduce noise in the trace. However, the way it's implemented (particularly lines 72-75) means that once you go over 10 calls, it will always be skipped. This also adds a fun situation where the following recursive code won't close any nvtx regions it opens:

```
void foo(void) {
    static int calls = 0;
    calls++;
    
    int skipstart = dr_hook_nvtx_map_start("foo");
    std::cout<<"In  call "<<calls<<" Skipped? "<<!skipstart<<std::endl;
    
    
    if (calls < 11) {
        foo();
    }
    
    int skipend = dr_hook_nvtx_map_stop();
    std::cout<<"Out call "<<calls<<" Skipped? "<<!skipend<<std::endl;
    
    calls--;
    return;
}
```

This gave the following output:
```
In  call 1 Skipped? 0
In  call 2 Skipped? 0
In  call 3 Skipped? 0
In  call 4 Skipped? 0
In  call 5 Skipped? 0
In  call 6 Skipped? 0
In  call 7 Skipped? 0
In  call 8 Skipped? 0
In  call 9 Skipped? 0
In  call 10 Skipped? 0
In  call 11 Skipped? 1
Out call 11 Skipped? 1
Out call 10 Skipped? 1
Out call 9 Skipped? 1
Out call 8 Skipped? 1
Out call 7 Skipped? 1
Out call 6 Skipped? 1
Out call 5 Skipped? 1
Out call 4 Skipped? 1
Out call 3 Skipped? 1
Out call 2 Skipped? 1
Out call 1 Skipped? 1
```

Am I correct in thinking this is not the intended behaviour and I should attempt to fix it in my patch?